schema/canonical/crc.go

```diff
@@ -46,12 +27,53 @@ func init() {
 	}
 }
 
-func AvroCRC64Fingerprint(data []byte) string {
+func AvroCRC64Fingerprint(schema []byte, writer io.Writer) error {
 	fp := EMPTY
-	for _, d := range data {
+	for _, d := range schema {
 		fp = (fp >> 8) ^ FP_TABLE[(fp^uint64(d))&0xff]
 	}
-	output := bytes.NewBuffer(make([]byte, 0, 8))
-	binary.Write(output, binary.LittleEndian, fp)
-	return hex.EncodeToString(output.Bytes())
+	err := binary.Write(writer, binary.LittleEndian, fp)
+	return err
+}
+
+func AvroCalcSchemaUID(schema string) []byte {
+	b := make([]byte, 0, 8)
+	output := bytes.NewBuffer(b)
+	ns := parser.NewNamespace(false)
+
+	s, err := ns.TypeForSchema([]byte(schema))
+	if err != nil {
+		panic("Invalid schema 1")
+	}
+	for _, def := range ns.Roots {
+		err = resolver.ResolveDefinition(def, ns.Definitions)
+		if err != nil {
+			panic("Invalid schema 2")
+		}
+	}
+	canonical, err := json.Marshal(CanonicalForm(s))
+	err = AvroCRC64Fingerprint(canonical, output)
+	if err != nil {
+		panic("Invalid schema 1")
+	}
+	return output.Bytes()
+}
+
+func AvroVersionHeader(writer io.Writer, header []byte) error {
+	fp := HeaderV1
+	err := binary.Write(writer, binary.LittleEndian, fp)
+	if err != nil {
+		return err
+	}
+	err = binary.Write(writer, binary.LittleEndian, header)
+	return err
+}
+
+func AvroConsumeHeader(r io.Reader) error {
+	buf := make([]byte, 10)
+	n, err := r.Read(buf)
+	if n < 10 || err != nil {
+		return errors.New("message length <2")
+	}
+	return nil
 }
```

Comment:
This file is all about CRC computation, I think it would make more sense to put the single-object encoding magic number and framing code in the soe package.

Reply:
done